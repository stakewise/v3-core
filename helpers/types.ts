import { BigNumberish } from 'ethers'
export type ThenArg<T> = T extends PromiseLike<infer U> ? U : T

export enum Networks {
  mainnet = 'mainnet',
  goerli = 'goerli',
}

export type NetworkConfig = {
  url: string
  chainId: number

  governor: string
  validatorsRegistry: string
  securityDeposit: BigNumberish

  // Keeper
  oracles: string[]
  rewardsMinOracles: number
  validatorsMinOracles: number
  rewardsDelay: number
  oraclesConfigIpfsHash: string
  maxAvgRewardPerSecond: BigNumberish

  // OsToken
  treasury: string
  osTokenFeePercent: number
  osTokenCapacity: BigNumberish
  osTokenName: string
  osTokenSymbol: string

  // OsTokenConfig
  redeemFromLtvPercent: number
  redeemToLtvPercent: number
  liqThresholdPercent: number
  liqBonusPercent: number
  ltvPercent: number

  // GenesisEthVault
  genesisVault: {
    admin: string
    poolEscrow: string
    stakedEthToken: string
    capacity: BigNumberish
    feePercent: number
  }
}
