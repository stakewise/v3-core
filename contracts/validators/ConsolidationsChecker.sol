// SPDX-License-Identifier: BUSL-1.1

pragma solidity ^0.8.22;

import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {IConsolidationsChecker} from '../interfaces/IConsolidationsChecker.sol';
import {Errors} from '../libraries/Errors.sol';
import {IKeeper} from '../interfaces/IKeeper.sol';

/**
 * @title ConsolidationsChecker
 * @author StakeWise
 * @notice Defines the functionality for checking signatures of oracles for validators consolidations
 */
contract ConsolidationsChecker is EIP712, IConsolidationsChecker {
  uint256 private constant _signatureLength = 65;
  bytes32 private constant _consolidationsCheckerTypeHash =
    keccak256('ConsolidationsChecker(address vault,bytes validators)');

  IKeeper private immutable _keeper;

  /**
   * @dev Constructor
   * @param keeper The address of the Keeper contract
   */
  constructor(address keeper) EIP712('ConsolidationsChecker', '1') {
    _keeper = IKeeper(keeper);
  }

  /// @inheritdoc IConsolidationsChecker
  function verifySignatures(
    address vault,
    bytes calldata validators,
    bytes calldata signatures
  ) external view override {
    if (!isValidSignatures(vault, validators, signatures)) {
      revert Errors.InvalidSignatures();
    }
  }

  /// @inheritdoc IConsolidationsChecker
  function isValidSignatures(
    address vault,
    bytes calldata validators,
    bytes calldata signatures
  ) public view override returns (bool) {
    return
      _isValidSignatures(
        _keeper.validatorsMinOracles(),
        keccak256(abi.encode(_consolidationsCheckerTypeHash, vault, keccak256(validators))),
        signatures
      );
  }

  /**
   * @notice Internal function for verifying oracles' signatures
   * @param requiredSignatures The number of signatures required for the verification to pass
   * @param message The message that was signed
   * @param signatures The concatenation of the oracles' signatures
   * @return True if the signatures are valid, otherwise false
   */
  function _isValidSignatures(
    uint256 requiredSignatures,
    bytes32 message,
    bytes calldata signatures
  ) private view returns (bool) {
    if (requiredSignatures == 0) {
      return false;
    }

    // check whether enough signatures
    unchecked {
      // cannot realistically overflow
      if (signatures.length < requiredSignatures * _signatureLength) {
        return false;
      }
    }

    bytes32 data = _hashTypedDataV4(message);
    address lastOracle;
    address currentOracle;
    uint256 startIndex;
    for (uint256 i = 0; i < requiredSignatures; i++) {
      unchecked {
        // cannot overflow as signatures.length is checked above
        currentOracle = ECDSA.recover(data, signatures[startIndex:startIndex + _signatureLength]);
      }
      // signatures must be sorted by oracles' addresses and not repeat
      if (currentOracle <= lastOracle || !_keeper.isOracle(currentOracle)) {
        return false;
      }

      // update last oracle
      lastOracle = currentOracle;

      unchecked {
        // cannot realistically overflow
        startIndex += _signatureLength;
      }
    }

    return true;
  }
}
