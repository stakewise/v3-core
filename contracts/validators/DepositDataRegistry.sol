// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {MerkleProof} from '@openzeppelin/contracts/utils/cryptography/MerkleProof.sol';
import {IDepositDataRegistry} from '../interfaces/IDepositDataRegistry.sol';
import {IKeeperValidators} from '../interfaces/IKeeperValidators.sol';
import {IVaultAdmin} from '../interfaces/IVaultAdmin.sol';
import {IVaultValidators} from '../interfaces/IVaultValidators.sol';
import {IVaultVersion} from '../interfaces/IVaultVersion.sol';
import {IVaultsRegistry} from '../interfaces/IVaultsRegistry.sol';
import {IVaultState} from '../interfaces/IVaultState.sol';
import {IKeeperRewards} from '../interfaces/IKeeperRewards.sol';
import {Errors} from '../libraries/Errors.sol';
import {Multicall} from '../base/Multicall.sol';

/**
 * @title DepositDataRegistry
 * @author StakeWise
 * @notice Defines the functionality for the Vault's deposit data management
 */
contract DepositDataRegistry is Multicall, IDepositDataRegistry {
  IVaultsRegistry private immutable _vaultsRegistry;

  /// @inheritdoc IDepositDataRegistry
  mapping(address => uint256) public override depositDataIndexes;

  /// @inheritdoc IDepositDataRegistry
  mapping(address => bytes32) public override depositDataRoots;

  mapping(address => address) private _depositDataManagers;
  mapping(address => bool) private _migrated;

  /**
   * @dev Modifier to check if the caller is a valid vault
   * @param vault The address of the vault
   */
  modifier onlyValidVault(address vault) {
    if (!_vaultsRegistry.vaults(vault) || IVaultVersion(vault).version() < 2) {
      revert Errors.InvalidVault();
    }
    _;
  }

  /**
   * @dev Constructor
   * @param vaultsRegistry The address of the vaults registry contract
   */
  constructor(address vaultsRegistry) {
    _vaultsRegistry = IVaultsRegistry(vaultsRegistry);
  }

  /// @inheritdoc IDepositDataRegistry
  function getDepositDataManager(address vault) public view override returns (address) {
    address depositDataManager = _depositDataManagers[vault];
    return depositDataManager == address(0) ? IVaultAdmin(vault).admin() : depositDataManager;
  }

  /// @inheritdoc IDepositDataRegistry
  function setDepositDataManager(
    address vault,
    address depositDataManager
  ) external override onlyValidVault(vault) {
    // only vault admin can set deposit data manager
    if (msg.sender != IVaultAdmin(vault).admin()) revert Errors.AccessDenied();

    // update deposit data manager
    _depositDataManagers[vault] = depositDataManager;
    emit DepositDataManagerUpdated(vault, depositDataManager);
  }

  /// @inheritdoc IDepositDataRegistry
  function setDepositDataRoot(
    address vault,
    bytes32 depositDataRoot
  ) external override onlyValidVault(vault) {
    if (msg.sender != getDepositDataManager(vault)) revert Errors.AccessDenied();
    if (depositDataRoots[vault] == depositDataRoot) revert Errors.ValueNotChanged();

    depositDataRoots[vault] = depositDataRoot;
    // reset validator index on every root update
    depositDataIndexes[vault] = 0;
    emit DepositDataRootUpdated(vault, depositDataRoot);
  }

  /// @inheritdoc IDepositDataRegistry
  function updateVaultState(
    address vault,
    IKeeperRewards.HarvestParams calldata harvestParams
  ) external override {
    IVaultState(vault).updateState(harvestParams);
  }

  /// @inheritdoc IDepositDataRegistry
  function registerValidator(
    address vault,
    IKeeperValidators.ApprovalParams calldata keeperParams,
    bytes32[] calldata proof
  ) external override onlyValidVault(vault) {
    // register validator
    IVaultValidators(vault).registerValidators(keeperParams, '');

    // SLOAD to memory
    uint256 currentIndex = depositDataIndexes[vault];
    bytes32 depositDataRoot = depositDataRoots[vault];

    // check matches merkle root and next validator index
    if (
      !MerkleProof.verifyCalldata(
        proof,
        depositDataRoot,
        keccak256(bytes.concat(keccak256(abi.encode(keeperParams.validators, currentIndex))))
      )
    ) {
      revert Errors.InvalidProof();
    }

    // increment index for the next validator
    unchecked {
      // cannot realistically overflow
      depositDataIndexes[vault] = currentIndex + 1;
    }
  }

  /// @inheritdoc IDepositDataRegistry
  function registerValidators(
    address vault,
    IKeeperValidators.ApprovalParams calldata keeperParams,
    uint256[] calldata indexes,
    bool[] calldata proofFlags,
    bytes32[] calldata proof
  ) external override onlyValidVault(vault) {
    // register validator
    IVaultValidators(vault).registerValidators(keeperParams, '');

    // SLOAD to memory
    uint256 currentIndex = depositDataIndexes[vault];
    bytes32 depositDataRoot = depositDataRoots[vault];

    // define leaves for multiproof
    uint256 validatorsCount = indexes.length;
    if (validatorsCount == 0) revert Errors.InvalidValidators();
    bytes32[] memory leaves = new bytes32[](validatorsCount);

    // calculate validator length
    uint256 validatorLength = keeperParams.validators.length / validatorsCount;
    if (validatorLength == 0) revert Errors.InvalidValidators();

    // calculate leaves
    {
      uint256 startIndex;
      uint256 endIndex;
      for (uint256 i = 0; i < validatorsCount; ) {
        endIndex += validatorLength;
        leaves[indexes[i]] = keccak256(
          bytes.concat(
            keccak256(abi.encode(keeperParams.validators[startIndex:endIndex], currentIndex))
          )
        );

        startIndex = endIndex;
        unchecked {
          // cannot realistically overflow
          ++currentIndex;
          ++i;
        }
      }
    }

    // check matches merkle root and next validator index
    if (!MerkleProof.multiProofVerifyCalldata(proof, proofFlags, depositDataRoot, leaves)) {
      revert Errors.InvalidProof();
    }

    // increment index for the next validator
    depositDataIndexes[vault] = currentIndex;
  }

  /// @inheritdoc IDepositDataRegistry
  function migrate(
    bytes32 depositDataRoot,
    uint256 validatorIndex,
    address depositDataManager
  ) external override onlyValidVault(msg.sender) {
    if (_migrated[msg.sender]) revert Errors.AccessDenied();

    depositDataRoots[msg.sender] = depositDataRoot;
    depositDataIndexes[msg.sender] = validatorIndex;
    _depositDataManagers[msg.sender] = depositDataManager;

    // only allow migration once
    _migrated[msg.sender] = true;
    emit DepositDataMigrated(msg.sender, depositDataRoot, validatorIndex, depositDataManager);
  }
}
