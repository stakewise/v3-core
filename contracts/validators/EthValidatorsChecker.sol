// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {ValidatorsChecker} from './ValidatorsChecker.sol';

/**
 * @title EthValidatorsChecker
 * @author StakeWise
 * @notice Defines functionality for checking validators registration on Ethereum
 */
contract EthValidatorsChecker is ValidatorsChecker {
  /**
   * @dev Constructor
   * @param validatorsRegistry The address of the beacon chain validators registry contract
   * @param keeper The address of the Keeper contract
   * @param vaultsRegistry The address of the VaultsRegistry contract
   * @param depositDataRegistry The address of the DepositDataRegistry contract
   */
  constructor(
    address validatorsRegistry,
    address keeper,
    address vaultsRegistry,
    address depositDataRegistry
  ) ValidatorsChecker(validatorsRegistry, keeper, vaultsRegistry, depositDataRegistry) {}

  /// @inheritdoc ValidatorsChecker
  function _depositAmount() internal pure override returns (uint256) {
    return 32 ether;
  }
}
