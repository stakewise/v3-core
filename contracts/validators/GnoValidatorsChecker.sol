// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {IValidatorsRegistry} from '../interfaces/IValidatorsRegistry.sol';
import {IKeeper} from '../interfaces/IKeeper.sol';
import {IDepositDataRegistry} from '../interfaces/IDepositDataRegistry.sol';
import {IVaultsRegistry} from '../interfaces/IVaultsRegistry.sol';
import {ValidatorsChecker} from './ValidatorsChecker.sol';

/**
 * @title GnoValidatorsChecker
 * @author StakeWise
 * @notice Defines Gnosis-specific settings for ValidatorsChecker contract
 */
contract GnoValidatorsChecker is ValidatorsChecker {
  /**
   * @dev Constructor
   * @param validatorsRegistry The address of the beacon chain validators registry contract
   * @param keeper The address of the Keeper contract
   * @param vaultsRegistry The address of the VaultsRegistry contract
   * @param depositDataRegistry The address of the DepositDataRegistry contract
   */
  constructor(
    address validatorsRegistry,
    address keeper,
    address vaultsRegistry,
    address depositDataRegistry
  )
    ValidatorsChecker(validatorsRegistry, keeper, vaultsRegistry, depositDataRegistry)
    EIP712('GnoValidatorsChecker', '1')
  {}

  function validatorsManagerSignatureTypeHash() internal pure override returns (bytes32) {
    return
      keccak256(
        'GnoValidatorsChecker(bytes32 validatorsRegistryRoot,address vault,bytes validators)'
      );
  }

  function depositAmount() internal pure override returns (uint256) {
    return 1 ether;
  }
}
