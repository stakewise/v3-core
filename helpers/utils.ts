import { Contract } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/src/types/runtime'

export async function deployContract(tx: any): Promise<Contract> {
  const result = await tx
  await result.deployTransaction.wait()
  return result
}

async function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function verify(
  hre: HardhatRuntimeEnvironment,
  address: string,
  constructorArgs: any,
  contractPath: string
) {
  if (!process.env.BLOCK_EXPLORER_KEY) {
    return
  }

  let count = 0
  const maxTries = 8
  // eslint-disable-next-line no-constant-condition
  while (true) {
    await delay(10000)
    try {
      console.log('Verifying contract at', address)
      await hre.run('verify:verify', {
        address,
        constructorArguments: constructorArgs,
        contract: contractPath,
      })
      break
    } catch (error) {
      if (String(error).includes('Already Verified')) {
        console.log(`Already verified contract at ${contractPath} at address ${address}`)
        break
      }
      if (++count == maxTries) {
        console.log(
          `Failed to verify contract at ${contractPath} at address ${address}, error: ${error}`
        )
        break
      }
      console.log(`Retrying... Retry #${count}, last error: ${error}`)
    }
  }
}
