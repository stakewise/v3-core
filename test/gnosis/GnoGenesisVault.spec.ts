import { ethers } from 'hardhat'
import { Contract, Wallet } from 'ethers'
import { loadFixture } from '@nomicfoundation/hardhat-toolbox/network-helpers'
import {
  ERC20Mock,
  GnoGenesisVault,
  Keeper,
  LegacyRewardTokenMock,
  PoolEscrowMock,
  DepositDataRegistry,
} from '../../typechain-types'
import { expect } from '../shared/expect'
import keccak256 from 'keccak256'
import {
  EXITING_ASSETS_MIN_DELAY,
  SECURITY_DEPOSIT,
  ZERO_ADDRESS,
  ZERO_BYTES32,
} from '../shared/constants'
import snapshotGasCost from '../shared/snapshotGasCost'
import { registerEthValidator } from '../shared/validators'
import { getHarvestParams, getRewardsRootProof, updateRewards } from '../shared/rewards'
import {
  extractDepositShares,
  extractExitPositionTicket,
  getBlockTimestamp,
  increaseTime,
} from '../shared/utils'
import { ThenArg } from '../../helpers/types'
import {
  collateralizeGnoVault,
  depositGno,
  gnoVaultFixture,
  setGnoWithdrawals,
} from '../shared/gnoFixtures'

describe('GnoGenesisVault', () => {
  const capacity = ethers.parseEther('1000000')
  const feePercent = 500
  const metadataIpfsHash = 'bafkreidivzimqfqtoqxkrpge6bjyhlvxqs3rhe73owtmdulaxr5do5in7u'
  let admin: Wallet, other: Wallet
  let vault: GnoGenesisVault, keeper: Keeper, validatorsRegistry: Contract, gnoToken: ERC20Mock
  let poolEscrow: PoolEscrowMock
  let rewardToken: LegacyRewardTokenMock, depositDataRegistry: DepositDataRegistry

  let createGenesisVault: ThenArg<ReturnType<typeof gnoVaultFixture>>['createGnoGenesisVault']

  async function acceptPoolEscrowOwnership() {
    await vault.connect(admin).acceptPoolEscrowOwnership()
  }

  async function collatGnoVault() {
    await collateralizeGnoVault(
      vault,
      gnoToken,
      keeper,
      depositDataRegistry,
      admin,
      validatorsRegistry
    )
  }

  beforeEach('deploy fixtures', async () => {
    ;[admin, other] = await (ethers as any).getSigners()
    const fixture = await loadFixture(gnoVaultFixture)
    keeper = fixture.keeper
    validatorsRegistry = fixture.validatorsRegistry
    gnoToken = fixture.gnoToken
    depositDataRegistry = fixture.depositDataRegistry
    ;[vault, rewardToken, poolEscrow] = await fixture.createGnoGenesisVault(admin, {
      capacity,
      feePercent,
      metadataIpfsHash,
    })
    createGenesisVault = fixture.createGnoGenesisVault
  })

  it('initializes correctly', async () => {
    await expect(vault.connect(admin).initialize(ZERO_BYTES32)).to.revertedWithCustomError(
      vault,
      'InvalidInitialization'
    )
    expect(await vault.capacity()).to.be.eq(capacity)

    // VaultAdmin
    const adminAddr = await admin.getAddress()

    // VaultVersion
    expect(await vault.version()).to.be.eq(3)
    expect(await vault.vaultId()).to.be.eq(`0x${keccak256('GnoGenesisVault').toString('hex')}`)

    // VaultFee
    expect(await vault.admin()).to.be.eq(adminAddr)
    expect(await vault.feeRecipient()).to.be.eq(adminAddr)
  })

  it('has version', async () => {
    expect(await vault.version()).to.eq(3)
  })

  describe('migrate', () => {
    it('fails from not rewardToken', async () => {
      await expect(
        vault.connect(admin).migrate(await admin.getAddress(), ethers.parseEther('1'))
      ).to.be.revertedWithCustomError(vault, 'AccessDenied')
    })

    it('fails when pool escrow ownership is not accepted', async () => {
      const [vault, rewardToken] = await createGenesisVault(admin, {
        capacity,
        feePercent,
        metadataIpfsHash,
      })
      const assets = ethers.parseEther('10')
      await expect(
        rewardToken.connect(other).migrate(other.address, assets, 0)
      ).to.be.revertedWithCustomError(vault, 'AccessDenied')
    })

    it('fails with zero receiver', async () => {
      await acceptPoolEscrowOwnership()
      await collatGnoVault()
      const assets = ethers.parseEther('1')
      await expect(
        rewardToken.connect(other).migrate(ZERO_ADDRESS, assets, assets)
      ).to.be.revertedWithCustomError(vault, 'ZeroAddress')
    })

    it('fails with zero assets', async () => {
      await acceptPoolEscrowOwnership()
      await collatGnoVault()
      await expect(
        rewardToken.connect(other).migrate(other.address, 0, 0)
      ).to.be.revertedWithCustomError(vault, 'InvalidAssets')
    })

    it('fails when not collateralized', async () => {
      const [vault, rewardToken] = await createGenesisVault(admin, {
        capacity,
        feePercent,
        metadataIpfsHash,
      })
      await vault.connect(admin).acceptPoolEscrowOwnership()
      const assets = ethers.parseEther('1')
      await expect(
        rewardToken.connect(other).migrate(other.address, assets, assets)
      ).to.be.revertedWithCustomError(vault, 'NotCollateralized')
    })

    it('fails when not harvested', async () => {
      await acceptPoolEscrowOwnership()
      await collatGnoVault()
      const reward = ethers.parseEther('5')
      const unlockedMevReward = 0n
      const vaultAddr = await vault.getAddress()
      const vaultReward = getHarvestParams(vaultAddr, reward, unlockedMevReward)
      await updateRewards(keeper, [vaultReward])
      await updateRewards(keeper, [
        getHarvestParams(vaultAddr, reward + ethers.parseEther('5'), unlockedMevReward),
      ])

      const holder = other
      const assets = ethers.parseEther('1')
      await expect(
        rewardToken.connect(holder).migrate(await holder.getAddress(), assets, assets)
      ).to.be.revertedWithCustomError(vault, 'NotHarvested')
    })

    it('migrates from rewardToken', async () => {
      await acceptPoolEscrowOwnership()
      await collatGnoVault()
      const assets = ethers.parseEther('10')
      const expectedShares = await vault.convertToShares(assets)

      const holder = other
      const holderAddr = await holder.getAddress()

      const receipt = await rewardToken.connect(holder).migrate(holderAddr, assets, 0)
      expect(await vault.getShares(holderAddr)).to.eq(expectedShares)

      await expect(receipt).to.emit(vault, 'Migrated').withArgs(holderAddr, assets, expectedShares)
      await snapshotGasCost(receipt)
    })
  })

  it('pulls withdrawals on claim exited assets', async () => {
    await acceptPoolEscrowOwnership()
    const assets = ethers.parseEther('1') - SECURITY_DEPOSIT
    let tx = await depositGno(vault, gnoToken, assets, other, other, ZERO_ADDRESS)
    const shares = await extractDepositShares(tx)
    expect(await vault.getShares(other.address)).to.eq(shares)

    // register validator
    await registerEthValidator(vault, keeper, depositDataRegistry, admin, validatorsRegistry)
    expect(await gnoToken.balanceOf(await vault.getAddress())).to.eq(0n)

    // enter exit queue
    tx = await vault.connect(other).enterExitQueue(shares, other.address)
    const positionTicket = await extractExitPositionTicket(tx)
    const timestamp = await getBlockTimestamp(tx)
    expect(await vault.getExitQueueIndex(positionTicket)).to.eq(-1)

    // withdrawals arrives
    await setGnoWithdrawals(validatorsRegistry, gnoToken, poolEscrow, assets)
    const exitQueueIndex = await vault.getExitQueueIndex(positionTicket)
    expect(exitQueueIndex).to.eq(0)
    expect(await vault.withdrawableAssets()).to.eq(0n)

    // claim exited assets
    await increaseTime(EXITING_ASSETS_MIN_DELAY)
    tx = await vault.connect(other).claimExitedAssets(positionTicket, timestamp, exitQueueIndex)
    await expect(tx)
      .to.emit(vault, 'ExitedAssetsClaimed')
      .withArgs(other.address, positionTicket, 0, assets)
    await snapshotGasCost(tx)
  })

  describe('update state', () => {
    let totalVaultAssets: bigint
    let totalLegacyAssets: bigint

    beforeEach(async () => {
      totalVaultAssets = ethers.parseEther('10')
      totalLegacyAssets = ethers.parseEther('5')
      await depositGno(
        vault,
        gnoToken,
        totalVaultAssets - SECURITY_DEPOSIT,
        other,
        other,
        ZERO_ADDRESS
      )
      await rewardToken.connect(other).setTotalStaked(totalLegacyAssets)
    })

    it('splits reward between rewardToken and vault', async () => {
      await acceptPoolEscrowOwnership()
      const reward = ethers.parseEther('30')
      const unlockedMevReward = 0n
      const expectedVaultDelta =
        (reward * totalVaultAssets) / (totalLegacyAssets + totalVaultAssets)
      const expectedLegacyDelta = reward - expectedVaultDelta
      const vaultReward = getHarvestParams(await vault.getAddress(), reward, unlockedMevReward)
      const rewardsTree = await updateRewards(keeper, [vaultReward])
      const proof = getRewardsRootProof(rewardsTree, vaultReward)
      const receipt = await vault.updateState({
        rewardsRoot: rewardsTree.root,
        reward: vaultReward.reward,
        unlockedMevReward: vaultReward.unlockedMevReward,
        proof,
      })

      expect(await rewardToken.totalAssets()).to.eq(totalLegacyAssets + expectedLegacyDelta)
      expect(await vault.totalAssets()).to.eq(totalVaultAssets + expectedVaultDelta)
      await snapshotGasCost(receipt)
    })

    it('skips updating legacy with zero total assets', async () => {
      await acceptPoolEscrowOwnership()
      await rewardToken.setTotalStaked(0n)
      await rewardToken.setTotalRewards(0n)
      await rewardToken.setTotalPenalty(0n)

      const reward = ethers.parseEther('5')
      const unlockedMevReward = 0n

      const vaultReward = getHarvestParams(await vault.getAddress(), reward, unlockedMevReward)
      const rewardsTree = await updateRewards(keeper, [vaultReward])
      const proof = getRewardsRootProof(rewardsTree, vaultReward)

      const totalLegacyAssetsBefore = await rewardToken.totalAssets()
      const totalVaultAssetsBefore = await vault.totalAssets()
      const receipt = await vault.updateState({
        rewardsRoot: rewardsTree.root,
        reward: vaultReward.reward,
        unlockedMevReward: vaultReward.unlockedMevReward,
        proof,
      })
      expect(await rewardToken.totalAssets()).to.eq(totalLegacyAssetsBefore)
      expect(await vault.totalAssets()).to.eq(totalVaultAssetsBefore + reward)
      await snapshotGasCost(receipt)
    })

    it('fails when pool escrow ownership not accepted', async () => {
      const [vault] = await createGenesisVault(admin, {
        capacity,
        feePercent,
        metadataIpfsHash,
      })
      const totalRewards = ethers.parseEther('30')
      const vaultReward = getHarvestParams(await vault.getAddress(), totalRewards, 0n)
      const rewardsTree = await updateRewards(keeper, [vaultReward])
      const proof = getRewardsRootProof(rewardsTree, vaultReward)
      await expect(
        vault.updateState({
          rewardsRoot: rewardsTree.root,
          reward: vaultReward.reward,
          unlockedMevReward: vaultReward.unlockedMevReward,
          proof,
        })
      ).to.be.revertedWithCustomError(vault, 'InvalidInitialHarvest')
    })

    it('fails with negative first update', async () => {
      const [vault] = await createGenesisVault(admin, {
        capacity,
        feePercent,
        metadataIpfsHash,
      })
      await vault.connect(admin).acceptPoolEscrowOwnership()
      const totalPenalty = ethers.parseEther('-5')
      const vaultReward = getHarvestParams(await vault.getAddress(), totalPenalty, 0n)
      const rewardsTree = await updateRewards(keeper, [vaultReward])
      const proof = getRewardsRootProof(rewardsTree, vaultReward)
      await expect(
        vault.updateState({
          rewardsRoot: rewardsTree.root,
          reward: vaultReward.reward,
          unlockedMevReward: vaultReward.unlockedMevReward,
          proof,
        })
      ).to.revertedWithCustomError(vault, 'InvalidInitialHarvest')
    })

    it('splits penalty between rewardToken and vault', async () => {
      await acceptPoolEscrowOwnership()
      await collatGnoVault()
      const reward = ethers.parseEther('-5')
      const unlockedMevReward = 0n
      const expectedVaultDelta =
        (reward * totalVaultAssets) / (totalLegacyAssets + totalVaultAssets)
      const expectedLegacyDelta = reward - expectedVaultDelta
      const vaultReward = getHarvestParams(await vault.getAddress(), reward, unlockedMevReward)
      const rewardsTree = await updateRewards(keeper, [vaultReward])
      const proof = getRewardsRootProof(rewardsTree, vaultReward)
      const receipt = await vault.updateState({
        rewardsRoot: rewardsTree.root,
        reward: vaultReward.reward,
        unlockedMevReward: vaultReward.unlockedMevReward,
        proof,
      })

      expect((await rewardToken.totalAssets()) - (await rewardToken.totalPenalty())).to.eq(
        totalLegacyAssets + expectedLegacyDelta + 1n // rounding error
      )
      expect(await vault.totalAssets()).to.eq(totalVaultAssets + expectedVaultDelta - 1n) // rounding error
      await snapshotGasCost(receipt)
    })

    it('deducts rewards on first state update', async () => {
      const [vault, rewardToken] = await createGenesisVault(admin, {
        capacity,
        feePercent,
        metadataIpfsHash,
      })
      await vault.connect(admin).acceptPoolEscrowOwnership()
      await depositGno(
        vault,
        gnoToken,
        totalVaultAssets - SECURITY_DEPOSIT,
        other,
        other,
        ZERO_ADDRESS
      )
      await rewardToken.connect(other).setTotalStaked(totalLegacyAssets)

      const totalRewards = ethers.parseEther('25')
      const legacyRewards = ethers.parseEther('5')
      await rewardToken.connect(other).setTotalRewards(legacyRewards)
      expect(await rewardToken.totalAssets()).to.eq(totalLegacyAssets + legacyRewards)
      expect(await rewardToken.totalRewards()).to.eq(legacyRewards)
      expect(await vault.totalAssets()).to.eq(totalVaultAssets)

      const expectedVaultDelta =
        ((totalRewards - legacyRewards) * totalVaultAssets) /
        (totalLegacyAssets + legacyRewards + totalVaultAssets)
      const expectedLegacyDelta = totalRewards - legacyRewards - expectedVaultDelta
      const vaultReward = getHarvestParams(await vault.getAddress(), totalRewards, 0n)
      const rewardsTree = await updateRewards(keeper, [vaultReward])
      const proof = getRewardsRootProof(rewardsTree, vaultReward)
      const receipt = await vault.updateState({
        rewardsRoot: rewardsTree.root,
        reward: vaultReward.reward,
        unlockedMevReward: vaultReward.unlockedMevReward,
        proof,
      })

      expect(await rewardToken.totalAssets()).to.eq(
        totalLegacyAssets + legacyRewards + expectedLegacyDelta
      )
      expect(await vault.totalAssets()).to.eq(totalVaultAssets + expectedVaultDelta)
      await snapshotGasCost(receipt)
    })
  })
})
