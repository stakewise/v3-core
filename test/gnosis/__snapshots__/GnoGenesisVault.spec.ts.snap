// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GnoGenesisVault migrate migrates from rewardToken 1`] = `
Object {
  "calldataByteLength": 100,
  "gasUsed": 79583,
}
`;

exports[`GnoGenesisVault pulls withdrawals on claim exited assets 1`] = `
Object {
  "calldataByteLength": 100,
  "gasUsed": 155206,
}
`;

exports[`GnoGenesisVault update state deducts rewards on first state update 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 148036,
}
`;

exports[`GnoGenesisVault update state skips updating legacy with zero total assets 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 143605,
}
`;

exports[`GnoGenesisVault update state splits penalty between rewardToken and vault 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 117411,
}
`;

exports[`GnoGenesisVault update state splits reward between rewardToken and vault 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 165136,
}
`;
